### PR TITLE
WIP: spike/nodes

### PR DESCRIPTION
**Pull request by jhrr**
After thinking a bit about the vague idea I suggested yesterday (a `Version` model and a `Node` model with recursive foreign-keys to itself), I realised it was a half-baked version of a much better one which had been nagging at me. We've been hard-coding the levels of the ATLAS schema for different texts, and now we are trying to find away to get around that, to get generic, and so we need to abstract somehow. Finding the right abstraction is always hard but I had noticed (as I'm sure we all have) that ATLAS data as we had been defining it so far is very much a tree but we hadn't really been exploiting that in full, or at least, the tree wasn't first-class. I had used a [Django SQL tree library](https://django-treebeard.readthedocs.io/en/latest/) before when working with Wagtail and it was really solid, and so I thought, why not go all in on the tree representation here. It actually turned out to be really nice and effortless, everything slots together and we get a tonne of things for free that we were previously doing a lot more labour to derive.

  - A single `Node` model all the way down, including the `Version` itself. We  can derive named, node levels to any depth required for each project by mapping against a `NODE_HIERARCHY` schema.
  - No need for any foreign keys **at all** we just insert nodes into the tree and that immediately gives us arbitrary structure. The order of insertion preserves precisely the default ATLAS order we are already using.
  - The fields `idx` and `position` can be derived from a node's position on the tree itself, there's no need to annotate them at ingestion.
  - 'Rank' comes for free: at what depth is a `Node` on the tree? If we don't count the `Version` URN as having a rank (ie. just the components making up a reference are ranked), then `Node.rank == Node.depth - 1`.
  - Deriving ancestors, children, etc. are all naturally and intuitively expressed by the [tree API](https://django-treebeard.readthedocs.io/en/latest/api.html). We get a full and extensible interface for tree filtering that we can customise as we see fit.
  - A `Node` knows if it has siblings (on either side) or if it is the first or last sibling, etc. this can help us with boundary annotations.
  - Ingestion becomes simple and generic. We define a URN hierarchy that maps the node types to node ranks when we destructure a URN reference into its component parts and instantiate them in order as `Node` objects along every branch of the tree. Use the tree API to insert each `Node` as a child of the `Node` above it.
  - We can use the tree filtering API as a template for an interface to override for alternative chunkings that are able to render new tree projections over a base data tree.
  - We can genericize the graphql schema endpoint over `Nodes`. All ranges can be derived with the same logic at any level on the hierarchy. As mentioned above, filtering ancestor, children and sibling payloads will become trivial (in the default chunking case).
  - Our ATLAS data is more or less read-only so the often higher write-costs of tree maintenance operations is not any concern.
  - Loads less code already.
  - Mandatory LOTR reference.

### CODE REVIEW

**Comment by jtauber** [comment]
One thing we should be mindful of (even if we punt it for a couple of sprints) is how this would work alternative chunking. It's entirely possible we just treat alternative chunking as a completely different tree with redundant content and just deal with that "denorm" at ingestion time.

So I'm inclined to go ahead with this approach and implement it for this sprint (possibly in both explore homer and the new server if we want to keep them in sync).

**Comment by jhrr** [comment]
Yeah, we did discuss chunking on the call yesterday so I'm sure Jake will
have some interesting thoughts to offer there.

**Comment by jhrr** [comment]
@jtauber I'd also note that, in terms of shape and hierarchy, the 'formal' tree that results from this approach is more or less isomorphic to the 'informal' tree that results from our current pattern, so I think the problem of how this would work with alternative chunking is more or less isomorphic to the problem of how we would work with alternative chunking using the current pattern. But here we get the benefits of wrapping things up in the tree API in order to go ahead and solve those problems.

**Comment by jacobwegner** [comment]
@jhrr this is a really great approach!

I actually got to a point during my spike today where I started to spin my wheels a bit on the best way to traverse the hierarchy, and it turns out trees are pretty good for that 😉 

(I also used `django-treebeard` on a wagtail site and am pretty impressed with a lot of what you get out of the box).

Hoping to give a fuller pass and fill in a few things that I discovered from my spike; hopefully within the next couple of hours.

If I don't make that deadline, maybe we can sync up on a call tomorrow, but with a few small caveats, I do think this is worth pursuing further.

**Comment by jacobwegner** [comment]
@jhrr beat me to replying to @jtauber about alternative chunking; I think there are still some things for us to mull about with regards to redundant content, but I think doing denorms would get us going, and then we could prove out with some performance testing to get a better idea on how some of those things might scale in terms of db bloat / query time

**Comment by jhrr** [comment]
@jacobwegner excellent news! follow up tomorrow is totally good for me.

**Comment by jacobwegner** [comment]
Going to leave a few separate comments here; have also opened [https://github.com/scaife-viewer/explorehomer-atlas/pull/15](https://github.com/scaife-viewer/explorehomer-atlas/pull/15) to demonstrate what I think we'll need to have something like the `passageLines` endpoint (but also supporting multiple heirarchies)

**Comment by jacobwegner** [comment]
@jhrr wrote:
> A single `Node` model all the way down, including the `Version` itself. We can derive named, node levels to any depth required for each project by mapping against a `NODE_HIERARCHY` schema.

Focusing on [CTS URNs](http://cite-architecture.github.io/ctsurn_spec/#overall-structure-of-a-cts-urn), we can decompose to:

`urn:cts`

`namespace`

`work`

`passage`

And the work can be decomposed to:

`textgroup`

`work`

`version`

`exemplar`

I think that those non-passage bits are consistent and we'll likely want to be able to query them from GraphQL and I think it'd be useful to have at the least an _alias_ into the tree so we can run queries against `versions`, etc.

See [this example](https://github.com/scaife-viewer/explorehomer-atlas/blob/0d4876a68f613b0503b88b3910a4c27a65f8ceb3/readhomer_atlas/library/schema.py#L114) in my spike PR.

The CTS URN spec also states that:

> The passage component is a hierarchy of one or more levels expressing a logical citation scheme applying to all versions of a text. A poem might be cited by the single unit of "poetic line." A prose work might be cited by a hierarchy such as "book/chapter/section/subsection." Passage references at any level of the text's citation hierarchy *may* identify either a single citable node or a range indicated by the first and last nodes of the range.

I think that similar to refsdecls, we'll want to declare within the version or exemplar how to identify the passage nodes (rather than assuming heterogeneous nodes at a site-level `NODE_HEIRARCHY`).

We might also benefit from splitting work / passage nodes up to not hit some of the [node limitations in treebeard](https://django-treebeard.readthedocs.io/en/latest/mp_tree.html#treebeard.mp_tree.MP_Node.steplen)

**Comment by jacobwegner** [comment]
@jhrr wrote:
> The fields `idx` and `position` can be derived from a node's position on the tree itself, there's no need to annotate them at ingestion.

As you'll see on my [spike PR](https://github.com/scaife-viewer/explorehomer-atlas/pull/15), I think that [`idx` should be calculated at ingestion](https://github.com/scaife-viewer/explorehomer-atlas/commit/0fb0f2016e81630eaff88290d6e42db4460c5190).  I tried a couple of things using some of the `next/prev` methods within treebeard, but being able to use the `idx` internally is performant in that we can compute some queries using math rather than having to do multiple round trips to the database or munge up the path.

I think position is fine to be a derived thing, and the `position` field was a relic from the `Book` / `Line` structure where we assumed that parts of the `ref` were integers.

We may also want to have a separate tree for passage nodes (versus work nodes), or otherwise we might need some special casing within the schema / GraphQL query layer to get the nodes at passage level _n_.

Reason being that if we wanted the "first" level of passage nodes on an exemplar, the depth in the tree would be different than passage nodes off of a version.

File: readhomer_atlas/library/importers.py [hunk 1]
```diff
@@ -3,67 +3,65 @@
...
+    refs = _destructure_ref(textpart_ref)
+    for idx, key in enumerate(refs):
+        parent = root_node if idx == 0 else nodes.get(refs[idx - 1])
+        node = nodes.get(key)
+        if node is None:
+            kind, ref = key
+            data = {"kind": kind, "urn": f"{root_node.urn}{ref}", "ref": ref}
+            if key == refs[-1]:
+                data.update({"text_content": tokens})
+            node = parent.add_child(**data)
```
Comment (jacobwegner) [inline readhomer_atlas/library/importers.py:44]: We might also look at using the `load_bulk` endpoint

https://django-treebeard.readthedocs.io/en/latest/api.html#treebeard.models.Node.load_bulk

**Comment by jacobwegner** [comment]
@jhrr wrote:

> We can genericize the graphql schema endpoint over `Nodes`. All ranges can be derived with the same logic at any level on the hierarchy. As mentioned above, filtering ancestor, children and sibling payloads will become trivial (in the default chunking case).

_and_:

> A `Node` knows if it has siblings (on either side) or if it is the first or last sibling, etc. this can help us with boundary annotations

I think at the "work" node level, we may be able to model ancestors / children / siblings from the tree, but I've also [experimented with driving some of those relationships from the urn and ref for nodes](https://github.com/scaife-viewer/explorehomer-atlas/blob/0d4876a68f613b0503b88b3910a4c27a65f8ceb3/readhomer_atlas/library/schema.py#L218):

I still think we're going to want a particular _passages_ endpoint to drive most of the content retrieval in the reader (where we can pass an arbitrary URN denoting a passage, but always ensure we get back the lowest text part nodes as the edges), e.g. passageTextParts [[code](https://github.com/scaife-viewer/explorehomer-atlas/blob/0d4876a68f613b0503b88b3910a4c27a65f8ceb3/readhomer_atlas/library/schema.py#L294) | [sample queries](https://github.com/scaife-viewer/explorehomer-atlas/blob/0d4876a68f613b0503b88b3910a4c27a65f8ceb3/README.md#sample-queries)]

We might want some sort of subclasses of `WorkNode` too internally; basic shortcuts to help resolve the `textgroup` as a prop, etc would be nice.

>  'Rank' comes for free: at what depth is a `Node` on the tree? If we don't count the `Version` URN as having a rank (ie. just the components making up a reference are ranked), then `Node.rank == Node.depth - 1`.

I agree with this, but depending on how we separate "work" parts vs "passage" parts, I really like exposing "rank" to denote 1 for book, 2 for line, etc.

**Pull request closed without merging**